fix(replace): Partial-version spec support

epage opened the pull request:
### What does this PR try to resolve?

#12614 changed package ID specs to allow fields in the version number to be optional.  This earliest branch with this change is `rust-1.74.0` (beta).  While @Eh2406 was investigating version metadata issues in #12772, problems with the partial version change were found
- `replace`s that specify version metadata were ignored **(fixed with this PR)**
- `replace`s that don't specify version metadata when the `Cargo.lock` contained a version metadata, it would previously be ignored (with a warning) but now match **(unchanged with this PR)**
  - When the version metadata in `Cargo.lock` differed from the overriding `Cargo.toml`, cargo would panic **(now an error in this PR)**

The questions before are
- Do we revert #12614 in `master` and `rust-1.74.0` or merge this PR into `master`
- If we merge this PR into `master`, do we cherry-pick this into `rust-1.74.0` or revert #12614, giving ourselves more time to find problems

### How should we test and review this PR?

The initial commit adds tests that pass as of #12614.  Prior to #12614, these tests would have warned that the `replace` was unused and failed because `bar::bar` didn't exist.  Each commit then changes the behavior (or not) and updates the corresponding test.

### Additional information

Eh2406 commented:
Thank you for the quick fix!

ehuss commented on src/cargo/core/resolver/dep_cache.rs:
```diff
@@ -173,11 +173,20 @@ impl<'a> RegistryQueryer<'a> {
+            );
+            if s.version() != summary.version() {
+                return Poll::Ready(Err(anyhow::anyhow!(
+                    "replacement specication `{}` matched {} and tried to override it with {}\n\
```
```suggestion
                    "replacement specification `{}` matched {} and tried to override it with {}\n\
```

ehuss commented on src/cargo/core/resolver/dep_cache.rs:
```diff
@@ -173,11 +173,20 @@ impl<'a> RegistryQueryer<'a> {
+            if s.version() != summary.version() {
+                return Poll::Ready(Err(anyhow::anyhow!(
+                    "replacement specication `{}` matched {} and tried to override it with {}\n\
+                     avoid matching unrelated packaged by being more specific",
```
```suggestion
                     avoid matching unrelated packages by being more specific",
```

ehuss commented on src/cargo/core/package_id_spec.rs:
```diff
@@ -176,8 +176,7 @@ impl PackageIdSpec {
         if let Some(ref v) = self.version {
-            let req = v.exact_req();
-            if !req.matches(package_id.version()) {
+            if !v.matches(package_id.version()) {
```
Can you update the `matching` unittest below to include the new behavior?

ehuss commented on src/cargo/util/semver_ext.rs:
```diff
@@ -182,16 +182,16 @@ impl PartialVersion {
-                pre: self.pre.as_ref().cloned().unwrap_or_default(),
-            }],
-        }
+    pub fn matches(&self, version: &Version) -> bool {
```
Can this method have a doc comment explaining that it matches build metadata, and that is a distinct decision that that caller needs to make whether or not that is what they want, since it is a subtle distinction?  (As compared to `Version::eq` versus `Version::cmp_precedence`, which is a subtle point that is easy to get wrong.)

ehuss reviewed (commented):
From what I can tell, this fixes much more than just `[replace]`, right?

For example, any command that takes a Package ID Spec (like `cargo build -p foo@1.2.3+123`), which on 1.73 (stable) will error if the metadata doesn't match, on beta it ignores it, and this PR fixes it to again error if it doesn't match, correct?

ehuss commented:
I'd be fine with backporting this to beta, since there is still about a month left. (Though reverting #12614 would also be fine, but I would lean towards just fixing it.)

epage commented:
> From what I can tell, this fixes much more than just `[replace]`, right?
> 
> For example, any command that takes a Package ID Spec (like `cargo build -p foo@1.2.3+123`), which on 1.73 (stable) will error if the metadata doesn't match, on beta it ignores it, and this PR fixes it to again error if it doesn't match, correct?

Yes, one of the bugs identifier and fixed in this (allowing different build metadata to still match) affects PackageIDSpecs everywhere.  I've updated the description to call this out.

In adding the unit tests, the fix for the above made it so `foo@1.2` will now match `foo@1.2.3-alpha.0`.  I'll call that out in the description as well so we can more explicitly decide if we want to special case pre-release or not.

epage commented:
We talked about this in the cargo team meeting.  We were fine with the change in
 behavior and with backporting this
- This required a cargo expert to detect the bug and was not found through users using nightly or beta
- With package id specs and replaces (deprecated) combined with build metadata and pre-release, the bast radius if a bug is fairly small and we can live with it broken in a release
- Though we were probably comfortable with `1.2.3` matching `1.2.3-alpha.0`, we decided to require opting into pre-releases, like version requirements, so we have more room for changing things if someone takes a broad view of pre-release and works to solve it.  d318ed4  has this change

Eh2406 commented:
Looks good, and if we missed something then new PR.
@bors r+

bors commented:
:pushpin: Commit 1e340663606ae0f16f44df353fa19e6474999dfd has been approved by `Eh2406`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=Eh2406 1e340663606ae0f16f44df353fa19e6474999dfd -->
<!-- homu: {"type":"Approved","sha":"1e340663606ae0f16f44df353fa19e6474999dfd","approver":"Eh2406","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 1e340663606ae0f16f44df353fa19e6474999dfd with merge 67271fd7a9584ca682196bca2183496bc5ea5ece...
<!-- homu: {"type":"BuildStarted","head_sha":"1e340663606ae0f16f44df353fa19e6474999dfd","merge_sha":"67271fd7a9584ca682196bca2183496bc5ea5ece"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/6563296035/job/17828806559)
Approved by: Eh2406
Pushing 67271fd7a9584ca682196bca2183496bc5ea5ece to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Eh2406","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/6563296035/job/17828806559"},"merge_sha":"67271fd7a9584ca682196bca2183496bc5ea5ece"} -->
